test/micro/org/openjdk/bench/vm/compiler/VectorBulkOperationsArray.java

```diff
@@ -0,0 +1,700 @@
+            "220","221","222","223","224","225","226","227","228","229",
+            "230","231","232","233","234","235","236","237","238","239",
+            "240","241","242","243","244","245","246","247","248","249",
+            "250","251","252","253","254","255","256","257","258","259",
+            "260","261","262","263","264","265","266","267","268","269",
+            "270","271","272","273","274","275","276","277","278","279",
+            "280","281","282","283","284","285","286","287","288","289",
+            "290","291","292","293","294","295","296","297","298","299",
+            "300"})
+    public static int NUM_ACCESS_ELEMENTS;
+
+    // Every array has two regions:
+    // - read region
+    // - write region
+    // We should make sure that the region is a multiple of 4k, so that the
+    // 4k-aliasing prevention trick can work.
+    //
+    // It would be ince to set REGION_SIZE statically, but we want to keep it rather small if possible,
+    // to avoid running out of cache. But it might be quite large if NUM_ACCESS_ELEMENTS is large.
+    public static int REGION_SIZE = -1024;
+    public static final int REGION_2_BYTE_OFFSET   = 1024 * 2; // prevent 4k-aliasing
+    public static final int REGION_2_SHORT_OFFSET  = REGION_2_BYTE_OFFSET / 2;
+    public static final int REGION_2_CHAR_OFFSET   = REGION_2_BYTE_OFFSET / 2;
+    public static final int REGION_2_INT_OFFSET    = REGION_2_BYTE_OFFSET / 4;
+    public static final int REGION_2_LONG_OFFSET   = REGION_2_BYTE_OFFSET / 8;
+    public static final int REGION_2_FLOAT_OFFSET  = REGION_2_BYTE_OFFSET / 4;
+    public static final int REGION_2_DOUBLE_OFFSET = REGION_2_BYTE_OFFSET / 8;
+    // For Objects, it could be 4 or 8 bytes. Dividing by 8 gives us something
+    // reasonable for both cases.
+    public static final int REGION_2_OBJECT_OFFSET = REGION_2_BYTE_OFFSET / 8;
+
+    // The arrays with the two regions each
+    private byte[] aB;
+    private short[] aS;
+    private char[] aC;
+    private int[] aI;
+    private long[] aL;
+    private float[] aF;
+    private double[] aD;
+
+    // Used when we need variable values in fill.
+    private byte varB = 42;
+    private short varS = 42;
+    private char varC = 42;
+    private int varI = 42;
+    private long varL = 42;
+    private float varF = 42;
+    private double varD = 42;
+
+    // Classes for Object arrays.
+    static class A {
+        int x;
+        A(int x) {
+            this.x = x;
+        }
+    }
+    static class B extends A {
+        int y;
+        B(int x, int y) {
+            super(x);
+            this.y = y;
+        }
+    }
+    private A[] aOA;
+    private B[] aOB;
+    private A varOA = new A(-1);
+    private B varOB = new B(-1, -1);
+
+    // Number of repetitions, to randomize the offsets.
+    public static final int REPETITIONS = 64 * 64;
+
+    @CompilerControl(CompilerControl.Mode.INLINE)
+    public static int offsetLoad(int i) { return i % 64; }
```

Comment:
it's a minor but `& 63`: since i is not proven to be positive, C2 doesn't strength reduce the modulus into the cheaper form (&).
you can mask `i` stripping out the negative bits too, and should work the same.

Reply:
Good idea! Though I think in all call sites of `offsetLoad` we know that `i` is positive, so I suspect that this is already enough.

Reply:
I can also do "shift and mask" for `offsetStore`. Though I would expect C2 to do the transformation already as well here too.